libs/@local/hash-backend-utils/src/machine-actors.ts

```diff
@@ -0,0 +1,297 @@
+
+      return getRoots(subgraph);
+    });
+
+  if (unexpectedEntities.length > 0) {
+    throw new Error(
+      `Critical: More than one machine entity with identifier ${identifier} found in the graph.`,
+    );
+  }
+
+  if (!machineEntity) {
+    throw new NotFoundError(
+      `Critical: No machine entity with identifier ${identifier} found in the graph.`,
+    );
+  }
+
+  return machineEntity.metadata.provenance.recordCreatedById;
+};
+
+/**
+ * Creates a Machine entity for an existing machine account that already has permissions in the specified web.
+ */
+export const createMachineActorEntity = async (
+  context: { graphApi: GraphApi },
+  {
+    identifier,
+    machineAccountId,
+    ownedById,
+    displayName,
+    shouldBeAbleToCreateMoreMachineEntities,
+    systemAccountId,
+  }: {
+    // A unique identifier for the machine actor
+    identifier: MachineActorIdentifier;
+    // An existing accountId for the machine actor, which will also be used to authenticate the request
+    machineAccountId: AccountId;
+    // The OwnedById of the web the actor's entity will belong to
+    ownedById: OwnedById;
+    // A display name for the machine actor, to display to users
+    displayName: string;
+    // Whether or not this machine should be able to create more machine entities after creating itself
+    shouldBeAbleToCreateMoreMachineEntities: boolean;
+    // The accountId of the system account, used to grant the machine actor permissions to instantiate system types
+    systemAccountId: AccountId;
+  },
+): Promise<EntityMetadata> => {
+  // Give the machine actor permissions to generate graph change notifications
+  await context.graphApi.modifyEntityTypeAuthorizationRelationships(
+    systemAccountId,
+    [
+      {
+        operation: "touch",
+        resource: systemEntityTypes.graphChangeNotification.entityTypeId,
+        relationAndSubject: {
+          subject: {
+            kind: "account",
+            subjectId: machineAccountId,
+          },
+          relation: "instantiator",
+        },
+      },
+    ],
+  );
+
+  // Give the machine actor permissions to instantiate its own entity (entities of type Machine)
+  await context.graphApi.modifyEntityTypeAuthorizationRelationships(
+    systemAccountId,
+    [
+      {
+        operation: "touch",
+        resource: systemEntityTypes.machine.entityTypeId,
+        relationAndSubject: {
+          subject: {
+            kind: "account",
+            subjectId: machineAccountId,
+          },
+          relation: "instantiator",
+        },
+      },
+    ],
+  );
+
+  const metadata = await context.graphApi
+    .createEntity(machineAccountId, {
+      draft: false,
+      entityTypeId: systemEntityTypes.machine.entityTypeId,
+      ownedById,
+      properties: {
+        [blockProtocolPropertyTypes.displayName.propertyTypeBaseUrl]:
+          displayName,
+        [systemPropertyTypes.machineIdentifier.propertyTypeBaseUrl]: identifier,
+      },
```

Comment:
Nit: if I recall correctly we wanted to start hardcoding the base URLs directly? Or did is this no longer the case. Appreciate this won't be as type-safe because we can't use `satisfies` as we discussed previously.
```suggestion
      properties: {
        "https://blockprotocol.org/@blockprotocol/types/property-type/display-name/":
          displayName,
        "https://hash.ai/@hash/types/property-type/machine-identifier/":
          identifier,
      } as MachineProperties,
```

Reply:
Done in https://github.com/hashintel/hash/pull/3714/commits/53ddf2e361b24932e24b2e9709dff4622c71ef3b